src/doomsday.rs

```diff
@@ -0,0 +1,123 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use async_std::sync::Mutex;
+pub use async_std::task::JoinHandle;
+use tracing::error;
+
+#[derive(Clone)]
+/// DoomsdayTimer will panic (`spawn_with_panic()`) or exit (`spawn_with_exit()`) if it is not
+/// `reset()` at least every `duration`
+pub struct DoomsdayTimer {
+    time_to_explode: Arc<Mutex<Instant>>,
+    duration: Duration,
+    defused: Arc<AtomicBool>,
+    aggressive_mode: bool,
+}
+
+impl DoomsdayTimer {
+    /// Spawn a new doomsday timer that will `panic()` if it explodes.
+    /// `awaiting` on the jh will panic if the `DoomsdayTimer` panicked
+    pub fn spawn_with_panic(duration: Duration) -> (Self, JoinHandle<()>) {
+        let s = Self {
+            time_to_explode: Arc::new(Mutex::new(Instant::now() + duration)),
+            duration,
+            defused: Default::default(),
+            aggressive_mode: false,
+        };
+
+        let cloned = s.clone();
+        let jh = async_std::task::spawn(async move {
+            cloned.main_loop().await;
+        });
+        (s, jh)
+    }
+
+    /// Spawn a new doomsday timer that will `exit(1)` if it explodes.
+    /// This will aggressively ensure the process does not survive
+    pub fn spawn_with_exit(duration: Duration) -> Self {
```

Comment:
This seems same as `spawn_with_panic`.  Why duplicate API?

Reply:
if you spawn a task that panics, it will not cause process to exit unless you await the jh that is returned by spawning. In other words, the panic might be swallowed. https://stackoverflow.com/questions/66790155/what-is-the-recommended-way-to-propagate-panics-in-tokio-tasks (this is about tokio, but same logic applies)

`spawn_with_exit` will cause process to exit, which is undesirable in other situations where panic is intended to be handled (unit tests use spawn_with_panic for this reason)

Reply:
codes are the same, except one returns JoinHandle, and the other doesn't.  So only need `spawn_with_panic`.  Caller just need to ignore handle

Reply:
Oh! I had a bug here and code was not supposed to be the same. 
spawn_with_exit should have been setting aggressive_mode to true!

Reply:
I thought spawn_with_exit vs spawn_with_panic was more clear than having `aggressive_mode` be part of constructor, but could be convinced other way

Reply:
I think exposing `aggresive_mode` will reduce code duplication

Reply:
works for me!